[CSS] Don't compile functional pseudo-classes like :is() with contextually invalid pseudo-element

mdubet opened the pull request:
#### 78f541198890a77f119dc63d8d80f376037cd781
<pre>
[CSS] Don&apos;t compile functional pseudo-classes like :is() with contextually invalid pseudo-element
<a href="https://bugs.webkit.org/show_bug.cgi?id=293230">https://bugs.webkit.org/show_bug.cgi?id=293230</a>
<a href="https://rdar.apple.com/151605806">rdar://151605806</a>

Reviewed by Simon Fraser.

This mostly reverts commit b9e946421ea6185c88ef9a881bbad23d4e8fb114 which is
buggy because it makes compiling :is(div::before) the same as compiling :is(div).
(the ::before fragment is &quot;optimized away&quot; as never matching,
but the div fragment is still being compiled and potentially matched)

Most of those cases (pseudo-element in functional pseudo-classes) are already prevented at parsing time,
but nesting resolution can create them after parsing.

The fix here is that whenever there is a pseudo-element (standalone or not) in :is(),
we fallback to the SelectorChecker.

* LayoutTests/TestExpectations:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: <a href="https://commits.webkit.org/296144@main">https://commits.webkit.org/296144@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f09df849ae0419c9c5c853162f5212d1771feb6e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27188 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17597 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/112722 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58042 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109470 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/27875 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/35689 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81610 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110435 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22082 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/96892 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/61994 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21520 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15032 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57486 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91450 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15064 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/115822 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34573 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/25476 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/90645 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/34948 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93147 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90391 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23046 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35306 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13084 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30329 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34494 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/40040 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34240 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37596 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/35901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on LayoutTests/imported/w3c/web-platform-tests/css/css-nesting/contextually-invalid-selectors.html:
```diff
@@ -31,6 +31,12 @@
     }
+
+    div::before {
+         & { /* `&` must not match the originating element */
```
Should this really be here and not in its own test?

Ahmad-S792 replied:
It is dedicated WPT test now, so I am just syncing it here - https://github.com/WebKit/WebKit/pull/46484 (Would appreciate review). @mdubet - you can rebase your PR post WPT sync and remove `TestExpectation`.

mdubet committed:
[CSS] Don't compile functional pseudo-classes like :is() with contextually invalid pseudo-element
https://bugs.webkit.org/show_bug.cgi?id=293230
rdar://151605806

Reviewed by Simon Fraser.

This mostly reverts commit b9e946421ea6185c88ef9a881bbad23d4e8fb114 which is
buggy because it makes compiling :is(div::before) the same as compiling :is(div).
(the ::before fragment is "optimized away" as never matching,
but the div fragment is still being compiled and potentially matched)

Most of those cases (pseudo-element in functional pseudo-classes) are already prevented at parsing time,
but nesting resolution can create them after parsing.

The fix here is that whenever there is a pseudo-element (standalone or not) in :is(),
we fallback to the SelectorChecker.

* LayoutTests/TestExpectations:
* Source/WebCore/css/SelectorChecker.cpp:
(WebCore::SelectorChecker::checkOne const):
* Source/WebCore/cssjit/SelectorCompiler.cpp:
(WebCore::SelectorCompiler::addPseudoClassType):

Canonical link: https://commits.webkit.org/296144@main
